src/thunderbots/software/geom/util.cpp

```diff
@@ -752,6 +752,59 @@ std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Ve
     return std::make_optional(intersection);
 }
 
+std::optional<Point> finiteLengthLineIntersection(const Vector &a, const Vector &b,
+                                                  const Vector &c, const Vector &d)
+{
+    std::optional<Point> intersection = lineIntersection(a, b, c, d);
+
+    if (intersection != std::nullopt)
+    {
+        // We know the point returned by lineIntersection() exists along the trajectory of
+        // one of the lines, now we must check that it is within the specified line points
+        // If point C exists on line A-B, DIST(AC) + DIST(CB) = DIST(AB)
+        // Calculated for lines a-b and c-d
+        if ((fabs(dist(a, *intersection)) + fabs(dist(*intersection, b)) ==
+             fabs(dist(a, b))) &&
+            fabs(dist(c, *intersection)) + fabs(dist(*intersection, d)) ==
+                fabs(dist(c, d)))
+        {
+            return intersection;
+        }
+    }
+
+    return std::nullopt;
+}
+
+std::optional<Point> velocityLineIntersection(const Vector element_velocity,
+                                              const Vector element_position,
+                                              const Point line_start,
+                                              const Point line_end,
+                                              const Rectangle boundry_area)
+{
+    // If the element is not within the boundries, return null
+    if (!boundry_area.containsPoint(element_position))
+    {
+        return std::nullopt;
+    }
+
+    // Calculate the closest edge of the boundry based on element speed and position
+    const double proximity_to_x_boundry = std::max(
+        (-boundry_area.width() / 2 - element_position.x()) / element_velocity.x(),
+        (boundry_area.width() / 2 - element_position.x()) / element_velocity.x());
+    const double proximity_to_y_boundry = std::max(
+        (-boundry_area.height() / 2 - element_position.y()) / element_velocity.y(),
+        (boundry_area.height() / 2 - element_position.y()) / element_velocity.y());
+
+    // Grab the boundry edge that is intersected first
+    const double proximity_to_closest_edge =
+        std::min(proximity_to_x_boundry, proximity_to_y_boundry);
+
+    const Point boundry_edge_intersection =
+        proximity_to_closest_edge * element_velocity + element_position;
+
+    return finiteLengthLineIntersection(line_start, line_end, element_position,
+                                        boundry_edge_intersection);
```

Comment:
I think we could shrink the changes here and reduce the amount of code we have to add if we make use of some of the functions we already have here. I took a quick stab at an implementation that fits with the revised function signature I suggested, let me know what you think.
```
std::tuple<std::optional<Point>, std::optional<Point>> raySegmentIntersect(const Ray& ray, const Segment& segment){
    auto isect = lineIntersection(ray.getRayStart(), ray.getDirection(),
                                  segment.getSegStart(), segment.getEnd());

    // Check that there was an intersection and it is within the ray and segment
    if (isect.has_value() && contains(ray, isect.value()) && contains(segment, isect.value())) {
        // Check if the ray and segment are parallel
        if(collinear(ray.getRayStart(), ray.getDirection(), segment.getSegStart())){
            // The start point of the overlapping segment between the ray and segment
            Point overlapping_seg_start = closestPointOnSeg(ray.getRayStart(), segment);

            // TODO: Calc the second point for the overlapping segment. This will probably
            //       need some logic to account for different overlaps (segment lies
            //       entirely within ray, both partially overlap, etc)

            return {overlapping_seg_start, overlapping_seg_end};
        }

        // They are not parallel, so just return the intersection
        return {isect, std::nullopt};
    }

    return {std::nullopt, std::nullopt};
}

```

Reply:
Implemented function.